sql/src/main/java/io/crate/analyze/relations/RelationNormalizer.java

```diff
@@ -0,0 +1,343 @@
+
+        private QuerySpec querySpec;
+
+        public Context(AnalysisMetaData analysisMetaData, List<Field> fields) {
+            this.analysisMetaData = analysisMetaData;
+            this.fields = fields;
+        }
+
+        public Collection<? extends Path> paths() {
+            return Collections2.transform(fields, new com.google.common.base.Function<Field, Path>() {
+                @Override
+                public Path apply(Field input) {
+                    return input.path();
+                }
+            });
+        }
+    }
+
+    private static class FieldReplacingVisitor extends ReplacingSymbolVisitor<Void> {
+
+        public static final FieldReplacingVisitor INSTANCE = new FieldReplacingVisitor(true);
+        public static final FieldRelationVisitor<Symbol> FIELD_RELATION_VISITOR = new FieldRelationVisitor(INSTANCE);
+
+        private FieldReplacingVisitor(boolean inPlace) {
+            super(inPlace);
+        }
+
+        @Override
+        public Symbol visitField(Field field, Void context) {
+            return FIELD_RELATION_VISITOR.process(field.relation(), field, field);
+        }
+    }
+
+    private static class AggregateFunctionReferenceVisitor extends SymbolVisitor<Void, Boolean> {
+
+        public static final AggregateFunctionReferenceVisitor INSTANCE = new AggregateFunctionReferenceVisitor();
+        public static final FieldRelationVisitor<Boolean> FIELD_RELATION_VISITOR = new FieldRelationVisitor(INSTANCE);
+
+        public static Boolean any(Symbol symbol) {
+            return INSTANCE.process(symbol, null);
+        }
+
+        @Override
+        protected Boolean visitSymbol(Symbol symbol, Void context) {
+            return false;
+        }
+
+        @Override
+        public Boolean visitFunction(Function symbol, Void context) {
+            if (FunctionInfo.Type.AGGREGATE.equals(symbol.info().type())) {
+                return true;
+            }
+            for (Symbol arg : symbol.arguments()) {
+                if (process(arg, context)) {
+                    return true;
+                }
+            }
+            return false;
+        }
+
+        @Override
+        public Boolean visitField(Field field, Void context) {
+            return FIELD_RELATION_VISITOR.process(field.relation(), field, false);
+        }
+    }
+
+    /**
+     * Visits an output symbol in a queried relation using the provided field index.
+     */
+    private static class FieldRelationVisitor<R> extends AnalyzedRelationVisitor<FieldRelationVisitorContext<R>, R> {
+
+        private final SymbolVisitor<Void, R> symbolVisitor;
+
+        public FieldRelationVisitor(SymbolVisitor<Void, R> symbolVisitor) {
+            this.symbolVisitor = symbolVisitor;
+        }
+
+        public R process(AnalyzedRelation relation, Field field, R defaultReturnValue) {
+            return process(relation, new FieldRelationVisitorContext<>(field.index(), defaultReturnValue));
```

Comment:
this would still work with the field only as context, since you could return null and return the default if process returns null.
using the field as context des not require to call .index on the field which is a computed value in most cases